### ISSUE TITLE
Weird behavior of pretrained models with CUDA

### ISSUE THREAD

**Issue by yoshitomo-matsubara** [InterDigitalInc/CompressAI#47]
Python: 3.6.9
CUDA: 11.2
compressai: 1.1.3
torch: 1.8.1

When replicating the RD curves in README for kodak dataset (`wget http://r0k.us/graphics/kodak/kodak/kodim{0,1,2}{0,1,2,3,4,5,6,7,8,9}.png` in `./kodak/`), I observed a weird behavior of pretrained models with CUDA.

# Looks like working well without CUDA
`python -m compressai.utils.eval_model pretrained ./kodak/ -a bmshj2018-hyperprior --metric mse --quality 5`

## metric = mse
```
Downloading: "https://compressai.s3.amazonaws.com/models/v1/bmshj2018-hyperprior-5-f8b614e1.pth.tar" to /home/yoshitom/.cache/torch/hub/checkpoints/bmshj2018-hyperprior-5-f8b614e1.pth.tar
100.0%
{
  "name": "bmshj2018-hyperprior",
  "description": "Inference (ans)",
  "results": {
    "psnr": [
      34.52624269077672
    ],
    "ms-ssim": [
      0.9835608204205831
    ],
    "bpp": [
      0.6686842176649305
    ],
    "encoding_time": [
      0.2404747505982717
    ],
    "decoding_time": [
      0.5095066924889883
    ]
  }
}
```
## metric = ms-ssim
`python -m compressai.utils.eval_model pretrained ./kodak/ -a bmshj2018-hyperprior --metric ms-ssim --quality 5`

```
Downloading: "https://compressai.s3.amazonaws.com/models/v1/bmshj2018-hyperprior-ms-ssim-5-c34afc8d.pth.tar" to /home/yoshitom/.cache/torch/hub/checkpoints/bmshj2018-hyperprior-ms-ssim-5-c34afc8d.pth.tar
100.0%
{
  "name": "bmshj2018-hyperprior",
  "description": "Inference (ans)",
  "results": {
    "psnr": [
      28.992422918554542
    ],
    "ms-ssim": [
      0.9866020356615385
    ],
    "bpp": [
      0.47353786892361116
    ],
    "encoding_time": [
      0.24171670277913412
    ],
    "decoding_time": [
      0.5283569494883219
    ]
  }
}
```

# PSNR and MS-SSIM are both NaN when using CUDA

`python -m compressai.utils.eval_model pretrained ./kodak/ -a bmshj2018-hyperprior --metric mse --quality 5 --cuda`

## metric = mse
```
Downloading: "https://compressai.s3.amazonaws.com/models/v1/bmshj2018-hyperprior-5-f8b614e1.pth.tar" to /home/yoshitom/.cache/torch/hub/checkpoints/bmshj2018-hyperprior-5-f8b614e1.pth.tar
100.0%
{
  "name": "bmshj2018-hyperprior",
  "description": "Inference (ans)",
  "results": {
    "psnr": [
      NaN
    ],
    "ms-ssim": [
      NaN
    ],
    "bpp": [
      0.6686876085069443
    ],
    "encoding_time": [
      0.034142365058263145
    ],
    "decoding_time": [
      0.025616129239400227
    ]
  }
}
```

`python -m compressai.utils.eval_model pretrained ./kodak/ -a bmshj2018-hyperprior --metric ms-ssim --quality 5 --cuda`

## metric = ms-ssim
```
Downloading: "https://compressai.s3.amazonaws.com/models/v1/bmshj2018-hyperprior-ms-ssim-5-c34afc8d.pth.tar" to /home/yoshitom/.cache/torch/hub/checkpoints/bmshj2018-hyperprior-ms-ssim-5-c34afc8d.pth.tar
100.0%
{
  "name": "bmshj2018-hyperprior",
  "description": "Inference (ans)",
  "results": {
    "psnr": [
      NaN
    ],
    "ms-ssim": [
      NaN
    ],
    "bpp": [
      0.47353786892361116
    ],
    "encoding_time": [
      0.03800355394681295
    ],
    "decoding_time": [
      0.029240707556406658
    ]
  }
}
```

I didn't check all the combinations (model, quality, metrics, with/without CUDA), but at least `bmshj2018-hyperprior ` with quality=8 (besides one with quality=5) also returned NaN when using CUDA (for both `mse` and `ms-ssim` checkpoints). There may be more checkpoints that face the same issue.

When I checked the output from a model (i.e., `out_dec["x_hat"]`), some value in the tensor is NaN when using CUDA and that must have caused this issue.

**Comment by jbegaint** [comment]
thanks for reporting, this is definitely strange, I'm going to take a look.

**Comment by jbegaint** [comment]
ok this looks like a regression with pytorch 1.8.1 and the entropy coding/decoding.

**Comment by jbegaint** [comment]
@yoshitomo-matsubara, so this is due to the nondeterministic behavior of some pytorch/cuda operations. We actually never encountered this issue when using the same pytorch version and hardware devices but looks like this changed in 1.8.1.

I'll push a fix tomorrow to enforce deterministic ops (`torch.backends.cudnn.deterministic = True`) where we need it.
We also have a models with a quantized hyperprior scales transform that would resolve this for all device types, we hope to release it soon.

**Comment by yoshitomo-matsubara** [comment]
@jbegaint Great to hear that you identified where to be fixed, thanks for the effort!
I'll be waiting for the release!

**Comment by jbegaint** [comment]
hey @yoshitomo-matsubara , could you try the current development version to check if it fixes the issues on your side?

**Comment by yoshitomo-matsubara** [comment]
Hi @jbegaint, Definitely, I think I can access the same environment tomorrow and let you know once it's done.
Is 1.1.4 dev available on pypi? or should I just fetch master branch and try it?

**Comment by jbegaint** [comment]
Thanks :-). You can fetch master, yes, it's not on pypi yet. Directly installing via `pip install git+https://github.com/InterDigitalInc/CompressAI/` should work also.

**Comment by yoshitomo-matsubara** [comment]
Hi @jbegaint, I fetched the master branch and tried the above configs for `bmshj2018-hyperprior` with/without CUDA.
The results with CUDA look same as those without CUDA. Thank you for the fix!

**Comment by jbegaint** [comment]
thanks for testing!
